### PR TITLE
[Paywalls V2] Implements Variables V2

### PR DESCRIPTION
**Pull request by JayShortway**
## Description

This is the Android equivalent of https://github.com/RevenueCat/purchases-ios/pull/4633 and https://github.com/RevenueCat/purchases-ios/pull/4740.

Decided to make a separate `VariableProcessorV2` just like on iOS. `VariableProcessor` would've needed too many risky changes otherwise. 

I tried to make this PR as small as possible (see https://github.com/RevenueCat/purchases-android/pull/2097 & https://github.com/RevenueCat/purchases-android/pull/2098), but it's still quite large. The biggest contributors to the diff are:
- `VariableProcessorV2`
- `TextComponentViewVariablesTests`
   
These also make up the meat of this PR unfortunately.

### CODE REVIEW

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/processed/VariableProcessorV2.kt [hunk 1]
```diff
@@ -0,0 +1,584 @@
...
+import com.revenuecat.purchases.ui.revenuecatui.helpers.Logger
+import java.text.DateFormat
+import java.util.Calendar
+import java.util.Currency
+import java.util.Date
+import java.util.Locale
+import kotlin.math.roundToInt
+
+@Suppress("TooManyFunctions")
+internal object VariableProcessorV2 {
```
Comment (tonidero) [inline ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/processed/VariableProcessorV2.kt:19]: ❤️

File: ui/revenuecatui/src/test/kotlin/com/revenuecat/purchases/ui/revenuecatui/components/text/TextComponentViewVariablesTests.kt [hunk 2]
```diff
@@ -0,0 +1,843 @@
...
+                    Args(
+                        packages = listOf(packageYearlyUsdTwoOffers),
+                        locale = "en_US",
+                        storefrontCountryCode = "US",
+                        variableLocalizations = variableLocalizationKeysForEnUs(),
+                    ),
+                    "\$"
+                )
+
+                Variable.PRODUCT_PERIODLY -> arrayOf(
```
Comment (tonidero) [inline ui/revenuecatui/src/test/kotlin/com/revenuecat/purchases/ui/revenuecatui/components/text/TextComponentViewVariablesTests.kt:172]: Should we add tests for lifetime/custom packages with some of these variables? I believe many of them just return null... But might be better to make sure with tests.

**Review by tonidero — APPROVED** [review]
Amazing work! This is looking great! Just a suggestion for some additional tests

File: ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt [hunk 3]
```diff
@@ -114,6 +116,27 @@ internal fun Offering.validatePaywallComponentsDataOrNull(
...
+        .mapError { nonEmptyListOf(it) }
+        .getOrElse { error -> return RcResult.Error(error) }
+
+    // Build a NonEmptyMap of variable localizations, ensuring that we always have the default localization as fallback.
+    val variableLocalizations = nonEmptyMapOf(
+        paywallComponents.data.defaultLocaleIdentifier to defaultVariableLocalization,
+        paywallComponents.uiConfig.localizations,
+    ).mapValues { (locale, map) ->
+        // We need to turn our NonEmptyMap<LocaleId, Map> into NonEmptyMap<LocaleId, NonEmptyMap>. If a certain locale
+        // has an empty Map, we add an AllLocalizationsMissing error for that locale to our list of errors.
```
[the remark this answers is not in the document]
Comment (JayShortway) [reply ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/helpers/OfferingToStateMapper.kt:134]: Yea you're right that we might loosen this up a little. Something we can fine tune later I think?
